README.md
Apache Zookeeper Cookbook
=========================

[![Cookbook Version](https://img.shields.io/cookbook/v/apache_zookeeper.svg)](https://community.opscode.com/cookbooks/apache_zookeeper)
[![Build Status](https://travis-ci.org/bbaugher/apache_zookeeper.svg?branch=master)](https://travis-ci.org/bbaugher/apache_zookeeper)

Installs and configures [Apache Zookeeper](http://zookeeper.apache.org/).

View the [Change Log](https://github.com/bbaugher/apache_zookeeper/blob/master/CHANGELOG.md) to see what has changed.

Supports
--------

 * CentOS
 * Ubuntu

Usage
-----

Using the default attributes will setup a single Zookeeper server in standalone mode.

If you are wanting to setup Zookeeper in a multi-server cluster make sure to fill out the
`node["apache_zookeeper"]["servers"]` like this,

    node["apache_zookeeper"]["servers"] = ["myzkhost1.com", "myzkhost2.com", myzkhost3.com"]

The array should include a value per server and can be any of the following values,

 * FQDN - `node['fqdn']`
 * Host Name - `node['hostname']`
 * Machine Name - `node['machinename']`
 * Any network interface - `node["network"]["interfaces"][..]`
 * EC2 Host Name - `node['ec2']['public_hostname']`
 * EC2 IP Address - `node['ec2']['public_ipv4']`

This array is used to configure/calculate the server ids for each zookeeper.

You can also provide the server ids yourself by including values for the attributes
`node["apache_zookeeper"]["zoo.cfg"]["server.[ID]"]`. If this is done you don't have to provide the
`node["apache_zookeeper"]["servers"]` attribute.

What does the installation look like
------------------------------------

By default the installation will look like,

    zkCli | /usr/bin/zkCli  - The Zookeeper cli binary command
    /opt/zookeeper/current/*        - All of Zookeeper's files (config, binaries, logs...)
    /var/opt/zookeeper/logs - Zookeeper log information
    /var/opt/zookeeper/data - Zookeeper data
    /etc/init.d/zookeeper   - An init.d script to start/stop zookeeper. You can use service
    				        zookeeper [start|stop|restart|status] instead

Unique Quorum and Leader Election Ports
---------------------------------------

It is possible to provide unique quorum and leader election ports in a few different ways.

    node["apache_zookeeper"]["servers"] = ["host1", "host2", "host3"]
    node["apache_zookeeper"]["follower_port"] = 2888
    node["apache_zookeeper"]["election_port"] = 3888

OR

    node["apache_zookeeper"]["servers"] = ["host1:2888:3888", "host2:2888:3888", "host3:2888:3888"]

OR

    node["apache_zookeeper"]["zoo.cfg"]["server.1"] = "host1:2888:3888"
    node["apache_zookeeper"]["zoo.cfg"]["server.2"] = "host2:2888:3888"
    node["apache_zookeeper"]["zoo.cfg"]["server.3"] = "host3:2888:3888"

Environment Variables
---------------------

Should note that the `zkServer.sh` and other various scripts provided by zookeeper use various environment variables to tweak
runtime settings. Here are some,

 * `ZOO_LOG_DIR` : Overwrites log4j `zookeeper.log.file`. Defaults to `.` if not set which is why we provide a default value for it to the `node["apache_zookeeper"]["log_dir"]` value.
 * `ZOO_LOG4J_PROP` : Overwrites log4j `zookeeper.root.logger`. Defaults to `'INFO, CONSOLE'` if not set which is why we provide a default value for it `'INFO,CONSOLE,ROLLINGFILE'`
 * `JMXDISABLE` : Disables jmx. Defaults to enabling JMX. To disable set to any value
 * `SERVER_JVMFLAGS` : JVM flags for the server process

Recipes
-------

 * `default`: Installs, configures and runs zookeeper as a service
 * `install`: Installs zookeeper
 * `configure`: Configures zookeeper
 * `service`: Runs zookeeper

Attributes
----------

 * `node["apache_zookeeper"]["install_java"]` : If you want to use the `java` cookbook to install java (default=`true`)
 * `node["apache_zookeeper"]["user"]` : The user that owns the Zookeeper installation (default="zookeeper")
 * `node["apache_zookeeper"]["group"]` : The group that owns the Zookeeper installation (default="zookeeper")
 * `node["apache_zookeeper"]["env_vars"]` : The environment variables set for the zookeeper user (default={"ZOO_LOG_DIR" => `node["apache_zookeeper"]["log_dir"]`, "ZOO_LOG4J_PROP" => "'INFO, CONSOLE, ROLLINGFILE'"})
 * `node["apache_zookeeper"]["servers"]` : The array of fqdn/hostnames/ips for the zookeeper servers in the cluster (default=[])
 * `node["apache_zookeeper"]["follower_port"]` : The port used by zookeeper followers (default=2888)
 * `node["apache_zookeeper"]["election_port"]` : The port used for zookeeper elections (default=3888)
 * `node["apache_zookeeper"]["version"]` : The version of Zookeeper to install (default="3.4.6")
 * `node["apache_zookeeper"]["mirror"]` : The URL to the mirror that hosts the zookeeper binary (default=`http://archive.apache.org/dist/zookeeper`)
 * `node["apache_zookeeper"]["binary_url"]` : The full binary url of Zookeeper. If you override this value make sure to provide a valid and up to date value for `node["apache_zookeeper"]["version"]` (default=`File.join node["apache_zookeeper"]["mirror"], "zookeeper-#{node["apache_zookeeper"]["version"]}", "zookeeper-#{node["apache_zookeeper"]["version"]}.tar.gz"`)
 * `node["apache_zookeeper"]["init_style"]` : The style of script to use for running the zookeeper service. Currently only supports `init`. (default = `init`)
 * `node["apache_zookeeper"]["install_dir"]` : The base directory Zookeeper should be installed into (default="/opt/zookeeper")
 * `node["apache_zookeeper"]["local_state_dir"]` : The base directory for all of Zookeepers local state (data, logs)
 * `node["apache_zookeeper"]["data_dir"]` : The data directory for Zookeeper (default=`"#{node["apache_zookeeper"]["local_state_dir"]}/data"`)
 * `node["apache_zookeeper"]["log_dir"]` : The log directory for Zookeeper (default=`"#{node["apache_zookeeper"]["local_state_dir"]}/logs"`)
 * `node["apache_zookeeper"]["bin_dir"]` : The bind directory for Zookeeper (default=`"#{node["apache_zookeeper"]["install_dir"]}/current/bin"`)
 * `node["apache_zookeeper"]["config_dir"]` : The config directory for Zookeeper (default=`"#{node["apache_zookeeper"]["install_dir"]}/current/config"`)
 * `node["apache_zookeeper"]["zoo.cfg"][*]` : The key/values set for the `zoo.cfg` config file (see attributes file for defaults)
 * `node["apache_zookeeper"]["log4j.properties"][*]` : The key/values set for the `log4j.properties` config file (see attributes file for defaults)
 * `node['apache_zookeeper']['install']['type']` : Type of zookeeper installation. Can be `source` or `package`
 * `node['apache_zookeeper']['package']['version']` : Default `nil`. Set if you need not the latest version.


CHANGELOG.md
Change Log
==========

[1.7.0 - 02-11-2020](https://github.com/bbaugher/apache_zookeeper/issues?milestone=22&state=closed)
---------------------------------------------------------------------------------------------------

  * [Enhancement] [Issue-57](https://github.com/bbaugher/apache_zookeeper/issues/57) : Fixed systemd deployment via ulimits and timeout

[1.6.0 - 06-11-2018](https://github.com/bbaugher/apache_zookeeper/issues?milestone=21&state=closed)
---------------------------------------------------------------------------------------------------

  * [Enhancement] [Issue-54](https://github.com/bbaugher/apache_zookeeper/issues/54) : Fix init.d/systemd to read configuration from `config_dir`

[1.5.0 - 06-08-2018](https://github.com/bbaugher/apache_zookeeper/issues?milestone=20&state=closed)
---------------------------------------------------------------------------------------------------

  * [Enhancement] [Issue-56](https://github.com/bbaugher/apache_zookeeper/issues/56) : Fix LSB init block to allow use of init script with systemd

[1.4.0 - 01-02-2018](https://github.com/bbaugher/apache_zookeeper/issues?milestone=19&state=closed)
---------------------------------------------------------------------------------------------------

  * [Enhancement] [Issue-52](https://github.com/bbaugher/apache_zookeeper/issues/52) : Fix for CentOS 7 while executing "apache_zookeeper::service_systemd" recipe.

[1.3.0 - 08-25-2017](https://github.com/bbaugher/apache_zookeeper/issues?milestone=18&state=closed)
---------------------------------------------------------------------------------------------------

  * [Enhancement] [Issue-50](https://github.com/bbaugher/apache_zookeeper/issues/50) : Correct README about log_dir and fix systemd

[1.2.0 - 08-04-2017](https://github.com/bbaugher/apache_zookeeper/issues?milestone=17&state=closed)
---------------------------------------------------------------------------------------------------

  * [Feature] [Issue-48](https://github.com/bbaugher/apache_zookeeper/issues/48) : Package installation support

[1.1.1 - 05-24-2017](https://github.com/bbaugher/apache_zookeeper/issues?milestone=16&state=closed)
---------------------------------------------------------------------------------------------------

  * [Bug] [Issue-47](https://github.com/bbaugher/apache_zookeeper/issues/47) : binary_url doesn't adhere to version or mirror

[1.1.0 - 05-24-2017](https://github.com/bbaugher/apache_zookeeper/issues?milestone=15&state=closed)
---------------------------------------------------------------------------------------------------

  * [Enhancement] [Issue-45](https://github.com/bbaugher/apache_zookeeper/issues/45) : Running service status with the correct user
  * [Bug] [Issue-40](https://github.com/bbaugher/apache_zookeeper/issues/40) : init.d scripts have the CONFIG path hardcoded.

[1.0.1 - 09-23-2016](https://github.com/bbaugher/apache_zookeeper/issues?milestone=14&state=closed)
---------------------------------------------------------------------------------------------------

  * [Bug] [Issue-43](https://github.com/bbaugher/apache_zookeeper/issues/43) : Readme references "binary_url", but the cookbook doesn't seem to actually have that

[1.0.0 - 12-11-2015](https://github.com/bbaugher/apache_zookeeper/issues?milestone=13&state=closed)
---------------------------------------------------------------------------------------------------

  * [Feature] [Issue-38](https://github.com/bbaugher/apache_zookeeper/issues/38) : Provide recipes for install/configure/service
  * [Feature] [Issue-37](https://github.com/bbaugher/apache_zookeeper/issues/37) : Support different init styles
  * [Enhancement] [Issue-36](https://github.com/bbaugher/apache_zookeeper/issues/36) : Remove xml-simple
  * [Bug] [Issue-27](https://github.com/bbaugher/apache_zookeeper/issues/27) : Package `daemon` is missing
  * [Enhancement] [Issue-25](https://github.com/bbaugher/apache_zookeeper/issues/25) : zookeeper as system user

[0.5.0 - 06-23-2015](https://github.com/bbaugher/apache_zookeeper/issues?milestone=12&state=closed)
---------------------------------------------------------------------------------------------------

  * [Enhancement] [Issue-24](https://github.com/bbaugher/apache_zookeeper/issues/24) : Fix up tests
  * [Enhancement] [Issue-23](https://github.com/bbaugher/apache_zookeeper/issues/23) : Recipe downgrades java to 1.6
  * [Enhancement] [Issue-20](https://github.com/bbaugher/apache_zookeeper/issues/20) : Add test zkcli works
  * [Enhancement] [Issue-13](https://github.com/bbaugher/apache_zookeeper/issues/13) : Use ulimit cookbook

[0.4.0 - 10-20-2014](https://github.com/bbaugher/apache_zookeeper/issues?milestone=11&state=closed)
---------------------------------------------------------------------------------------------------

  * [Enhancement] [Issue-19](https://github.com/bbaugher/apache_zookeeper/issues/19) : Documentation update for default ports
  * [Enhancement] [Issue-18](https://github.com/bbaugher/apache_zookeeper/issues/18) : Match any IP address, not just the default

[0.3.2 - 09-09-2014](https://github.com/bbaugher/apache_zookeeper/issues?milestone=10&state=closed)
---------------------------------------------------------------------------------------------------

  * [Bug] [Issue-16](https://github.com/bbaugher/apache_zookeeper/issues/16) : Version fix

[0.3.1 - 09-09-2014](https://github.com/bbaugher/apache_zookeeper/issues?milestone=9&state=closed)
--------------------------------------------------------------------------------------------------

  * [Bug] [Issue-14](https://github.com/bbaugher/apache_zookeeper/issues/14) : `attributes?` not a method -- affects ec2 provisioning

[0.3.0 - 08-21-2014](https://github.com/bbaugher/apache_zookeeper/issues?milestone=6&state=closed)
--------------------------------------------------------------------------------------------------


attributes/default.rb
# encoding: UTF-8
# Cookbook Name: apache_zookeeper
# Attributes:: default
#
# Apache Zookeeper global attributes

#################  Installation attributes ###########################
default['apache_zookeeper']['install_java'] = true
default['apache_zookeeper']['install']['type'] = 'source'

default['apache_zookeeper']['install_dir'] = '/opt/zookeeper'
default['apache_zookeeper']['version'] = '3.4.10'
default['apache_zookeeper']['mirror'] = 'http://archive.apache.org/dist/zookeeper'

# User/group creation and limits
default['apache_zookeeper']['service_timeout'] = 300 # 5 minutes
default['apache_zookeeper']['user'] = 'zookeeper'
default['apache_zookeeper']['group'] = 'zookeeper'
default['apache_zookeeper']['nfiles'] = 32768
default['ulimit']['users'][node['apache_zookeeper']['user']]['filehandle_limit'] = default['apache_zookeeper']['nfiles'] # ~FC047
default['ulimit']['users'][node['apache_zookeeper']['user']]['process_limit'] = 1024 # ~FC047

################ Package Attributes #############################
default['apache_zookeeper']['package']['version'] = nil

################ Configuration Attributes #############################

default['apache_zookeeper']['local_state_dir'] = '/var/opt/zookeeper'

# This are used to configure a cluster
default['apache_zookeeper']['servers'] = []
default['apache_zookeeper']['follower_port'] = 2888
default['apache_zookeeper']['election_port'] = 3888

############## Service Attributes #####################################
default['apache_zookeeper']['init_style'] = 'init'

################# Template attributes ################################
default['apache_zookeeper']['env_vars']['ZOO_LOG4J_PROP'] = 'INFO,ROLLINGFILE'

# Zookeeper configuration options
default['apache_zookeeper']['zoo.cfg']['clientPort'] = 2181
default['apache_zookeeper']['zoo.cfg']['tickTime'] = 2000
default['apache_zookeeper']['zoo.cfg']['autopurge.purgeInterval'] = 24
default['apache_zookeeper']['zoo.cfg']['initLimit'] = 10
default['apache_zookeeper']['zoo.cfg']['syncLimit'] = 5

# Settings from default zookeeper installation
default['apache_zookeeper']['log4j.properties']['zookeeper.root.logger'] = 'CONSOLE,ROLLINGFILE'
default['apache_zookeeper']['log4j.properties']['zookeeper.console.threshold'] = 'INFO'

default['apache_zookeeper']['log4j.properties']['zookeeper.log.file'] = 'zookeeper.log'
default['apache_zookeeper']['log4j.properties']['zookeeper.log.threshold'] = 'INFO'
default['apache_zookeeper']['log4j.properties']['zookeeper.tracelog.file'] = 'zookeeper_trace.log'
default['apache_zookeeper']['log4j.properties']['log4j.rootLogger'] = '${zookeeper.root.logger}'
default['apache_zookeeper']['log4j.properties']['log4j.appender.CONSOLE'] = 'org.apache.log4j.ConsoleAppender'
default['apache_zookeeper']['log4j.properties']['log4j.appender.CONSOLE.Threshold'] = '${zookeeper.console.threshold}'
default['apache_zookeeper']['log4j.properties']['log4j.appender.CONSOLE.layout'] = 'org.apache.log4j.PatternLayout'
default['apache_zookeeper']['log4j.properties']['log4j.appender.CONSOLE.layout.ConversionPattern'] = '%d{ISO8601} [myid:%X{myid}] - %-5p [%t:%C{1}@%L] - %m%n'
default['apache_zookeeper']['log4j.properties']['log4j.appender.ROLLINGFILE'] = 'org.apache.log4j.RollingFileAppender'
default['apache_zookeeper']['log4j.properties']['log4j.appender.ROLLINGFILE.Threshold'] = '${zookeeper.log.threshold}'
default['apache_zookeeper']['log4j.properties']['log4j.appender.ROLLINGFILE.File'] = '${zookeeper.log.dir}/${zookeeper.log.file}'
default['apache_zookeeper']['log4j.properties']['log4j.appender.ROLLINGFILE.MaxFileSize'] = '10MB'
default['apache_zookeeper']['log4j.properties']['log4j.appender.ROLLINGFILE.MaxBackupIndex'] = '10'
default['apache_zookeeper']['log4j.properties']['log4j.appender.ROLLINGFILE.layout'] = 'org.apache.log4j.PatternLayout'
default['apache_zookeeper']['log4j.properties']['log4j.appender.ROLLINGFILE.layout.ConversionPattern'] = '%d{ISO8601} [myid:%X{myid}] - %-5p [%t:%C{1}@%L] - %m%n'
default['apache_zookeeper']['log4j.properties']['log4j.appender.TRACEFILE'] = 'org.apache.log4j.FileAppender'
default['apache_zookeeper']['log4j.properties']['log4j.appender.TRACEFILE.Threshold'] = 'TRACE'
default['apache_zookeeper']['log4j.properties']['log4j.appender.TRACEFILE.File'] = '${zookeeper.log.dir}/${zookeeper.tracelog.file}'
default['apache_zookeeper']['log4j.properties']['log4j.appender.TRACEFILE.layout'] = 'org.apache.log4j.PatternLayout'
default['apache_zookeeper']['log4j.properties']['log4j.appender.TRACEFILE.layout.ConversionPattern'] = '%d{ISO8601} [myid:%X{myid}] - %-5p [%t:%C{1}@%L][%x] - %m%n'
